server/core/src/main/java/org/infinispan/server/iteration/DefaultIterationManager.java

```diff
@@ -151,7 +151,10 @@ public IterationState start(AdvancedCache cache, BitSet segments, String filterC
       DataConversion valueDataConversion = cache.getValueDataConversion();
       Function<Object, Object> unmarshaller = p -> encoderRegistry.convert(p, requestValueType, APPLICATION_OBJECT);
 
-      MediaType storageMediaType = cache.getValueDataConversion().getStorageMediaType();
+      // The iterator converts from the request type to the storage type. The filter will receive the data already converted.
+      MediaType storageMediaType = valueDataConversion.getRequestMediaType() != null && valueDataConversion.getRequestMediaType() != MediaType.APPLICATION_UNKNOWN
```

Comment:
Hrmm... this filter stuff looks wrong. We seem to always be applying the value media type to do the conversion and don't pay attention to the key media type at all... am I missing something?

Reply:
Yeah. I noticed this as well. This method only handles the MediaType for values. If a conversion happens, the key is converted based on the value type. Let me address your other comment, and I'll mix key/value types.